regcore/db/django_models.py

```diff
@@ -1,46 +1,86 @@
-                'node_type': reg.node_type,
-                'children': reg.children
-            }
-            if reg.title:
-                as_dict['title'] = reg.title
-            return as_dict
-        except ObjectDoesNotExist:
+        regs = Regulation.objects.filter(
+            version=version,
+            label_string=label,
+        ).get_descendants(
+            include_self=True,
+        )
+        regs = list(regs.all())
+        if not regs:
             return None
-
-    def _transform(self, reg, version):
+        tree = get_tree(regs)
+        return self.serialize(regs[0], tree)
+
+    def serialize(self, reg, tree):
+        ret = {
+            'label': reg.label_string.split('-'),
+            'text': reg.text,
+            'node_type': reg.node_type,
+            'children': [
+                self.serialize(child, tree)
+                for child in tree.get(reg.id, [])
+            ],
+        }
+        if reg.title:
+            ret['title'] = reg.title
+        return ret
+
+    def _transform(self, reg, version, parent=None):
         """Create the Django object"""
-        return Regulation(version=version,
-                          label_string='-'.join(reg['label']),
-                          text=reg['text'],
-                          title=reg.get('title', ''),
-                          node_type=reg['node_type'],
-                          root=(len(reg['label']) == 1),
-                          children=reg['children'])
+        return Regulation(
+            id=self._get_id(reg, version),
+            parent_id=self._get_id(parent, version) if parent else None,
+            version=version,
+            label_string='-'.join(reg['label']),
+            text=reg['text'],
+            title=reg.get('title', ''),
+            node_type=reg['node_type'],
+            root=(len(reg['label']) == 1),
+        )
+
+    def _get_id(self, reg, version):
+        return '{}-{}'.format(version, '-'.join(reg['label']))
 
     def bulk_put(self, regs, version, root_label):
         """Store all reg objects"""
         # This does not handle subparts. Ignoring that for now
         Regulation.objects.filter(version=version,
                                   label_string__startswith=root_label).delete()
-        Regulation.objects.bulk_create(
-            [self._transform(r, version) for r in regs], batch_size=100)
+
+        rows = [self._transform(reg, version, parent) for reg, parent in regs]
+
+        # Save root node to establish `tree_id`
+        rows[0].save()
+
+        # Bulk save children with placeholder tree values
+        for row in rows[1:]:
+            row.tree_id = rows[0].tree_id
+            row.level, row.lft, row.rght = 1, 1, 1
+        with Regulation.objects.disable_mptt_updates():
+            Regulation.objects.bulk_create(rows[1:], batch_size=50)
+
+        # Rebuild partial tree from root node
+        Regulation.objects.partial_rebuild(rows[0].tree_id)
```

Comment:
Re. workaround: yes, exactly. Since bulk updates bypass a lot of the normal behavior (calling `save`, emitting events, updating saved objects, etc.), django-mptt doesn't handle them for free. The recommendation in the docs and issues is to disable mptt updates, then explicitly rebuild the full or partial tree after bulk operations. I haven't gotten too far into the mptt internals yet, but as a quick explanation, none of this happens on `bulk_create`: https://github.com/django-mptt/django-mptt/blob/cf857138b2cb5eebc465c912c86e804ed975dd99/mptt/models.py#L807-L996.

Does this make sense? I may still be missing something.